src/ray/core_worker/store_provider/memory_store/memory_store.cc

```diff
@@ -0,0 +1,230 @@
+                                        std::vector<std::shared_ptr<RayObject>> *results,
+                                        bool is_get) {
+  (*results).resize(object_ids.size(), nullptr);
+  std::vector<ObjectID> remaining_ids;
+
+  std::shared_ptr<GetOrWaitRequest> get_request;
+
+  {
+    std::unique_lock<std::mutex> lock(lock_);
+    // Check for existing objects and see if this get request can be fullfilled.
+    for (int i = 0; i < object_ids.size(); i++) {
+      const auto &object_id = object_ids[i];
+      auto iter = objects_.find(object_id);
+      if (iter != objects_.end()) {
+        (*results)[i] = iter->second;
+        if (is_get) {
+          objects_.erase(object_id);
+        }
+      } else {
+        remaining_ids.emplace_back(object_id);
+      }
+    }
+
+    // Return if all the objects are obtained.
+    if (remaining_ids.empty()) {
+      return Status::OK();
+    }
+
+    // Otherwise, create a GetOrWaitRequest to track remaining objects.
+    get_request = std::make_shared<GetOrWaitRequest>(remaining_ids, is_get);
+    for (const auto &object_id : remaining_ids) {
+      object_get_requests_[object_id].push_back(get_request);
+    }
+  }
+
+  // Wait for remaining objects (or timeout).
+  get_request->Wait(timeout_ms);
+
+  {
+    std::unique_lock<std::mutex> lock(lock_);
+    // Populate results.
+    for (int i = 0; i < object_ids.size(); i++) {
+      const auto &object_id = object_ids[i];
+      if ((*results)[i] == nullptr) {
+        (*results)[i] = get_request->Get(object_id);
+      }
+    }
+
+    // Remove get rquest.
+    for (const auto &object_id : get_request->ObjectIds()) {
+      auto object_request_iter = object_get_requests_.find(object_id);
+      if (object_request_iter != object_get_requests_.end()) {
+        auto &get_requests = object_request_iter->second;
+        // Erase get_req from the vector.
+        auto it = std::find(get_requests.begin(), get_requests.end(), get_request);
+        if (it != get_requests.end()) {
+          get_requests.erase(it);
+          // If the vector is empty, remove the object ID from the map.
+          if (get_requests.empty()) {
+            object_get_requests_.erase(object_request_iter);
+          }
+        }
+      }
+    }
+  }
+
+  return Status::OK();
+}
+
+Status CoreWorkerMemoryStore::Get(const std::vector<ObjectID> &object_ids,
+                                  int64_t timeout_ms,
+                                  std::vector<std::shared_ptr<RayObject>> *results) {
+  return GetOrWait(object_ids, timeout_ms, results, /* is_get */ true);
+}
+
+Status CoreWorkerMemoryStore::Wait(const std::vector<ObjectID> &object_ids,
+                                   int num_objects, int64_t timeout_ms,
+                                   std::vector<bool> *results) {
+  if (num_objects != object_ids.size()) {
+    return Status::Invalid("num_objects should equal to number of items in object_ids");
```

Comment:
Is this limitation temporary?

Reply:
Thanks for catching this. The `num_objects` param should be removed for store, it's a copy-n-paste error. At store provider level it requires the numbers to match, while later when integrating this into core_worker, I'll implement a while loop inside the `StoreInterface` so that the user interface would actually support early return.